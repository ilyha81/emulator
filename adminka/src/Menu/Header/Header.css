.Header {
    grid-area: header;
    width: 100%;
    height: 100%;
    background: grey;
}

.logo {
    margin-top: -5px;
    width: auto;
    height: 80px;
    pointer-events: none;
    filter: sepia(0.5);
    transition: filter 0.2s;
}
.logo-rotation {
    animation: App-logo-spin infinite 5s ease-in-out alternate-reverse;
    filter: sepia(0);
}

@keyframes App-logo-spin {
    from {
        transform: rotate(0deg);
    }
    to {
        transform: rotate(360deg);
    }
}