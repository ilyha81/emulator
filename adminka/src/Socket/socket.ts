import io from "socket.io-client";
import {BehaviorSubject} from "rxjs";

/**
 * Типы сообщений.
 */
export enum ESocketMsg {
    MESSAGE = "MESSAGE",
    SHOW_DATA = "SHOW_DATA",
    TEST = "TEST",
}

export type TSocketMessageCB = (data: string) => void;

/**
 * Функция отрисовки данных в соответствующей секции
 * @param data Данные для отрисовки
 */
const drawData = (data: string) => {
    const section = document.getElementById('show-data');
    console.log(data);
    section && (section.innerText = JSON.stringify(data));
};

/**
 * Сам коннект Сокета.
 */
const socket = io.connect('http://localhost:8000', {transports: ['websocket', 'polling']});

/**
 * Создаем подписчикjd.
 */
const MessageSubscriber = new BehaviorSubject<string>("");
const OperationsSubscriber = new BehaviorSubject<string>("");

/**
 * Создаем рассылку.
 */
socket.on(ESocketMsg.MESSAGE, (data: string) => {
    console.warn("SCOKET: incomming message = " + ESocketMsg.MESSAGE);
    MessageSubscriber.next(data)
});
socket.on(ESocketMsg.SHOW_DATA, (data: string) => {
    console.warn("SCOKET: incomming message = " + ESocketMsg.SHOW_DATA);
    OperationsSubscriber.next(data)
});

export const SocketAPI = {
    /**
     * Создание тестовой операции.
     */
    emitTestOperation: () => {
        socket.emit(ESocketMsg.TEST, {
            documents: [{id: 'asd1', docType: 'RPP'}, {id: 'asd2', docType: 'RPP'}],
            cryptoprofileId: "TEST_PROFILE"
        })
    },

    /**
     * Подписка на Получение сообщения.
     * @param {string} cb Колбэк
     */
    subscribeOnHello: (cb: TSocketMessageCB) => MessageSubscriber.subscribe(cb),

    /**
     * Подписка на Получение операций.
     * @param {string} cb Колбэк.
     */
    subscribeOnOperations: (cb: TSocketMessageCB) => OperationsSubscriber.subscribe(cb),

};

