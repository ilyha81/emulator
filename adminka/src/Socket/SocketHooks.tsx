import { useState, useEffect } from 'react';
import {ESocketConnection, SocketAPI} from './socket';

/**
 * Hook subscribes on Socket connection status.
 */
export const useSocketConnectionStatus = ():ESocketConnection  => {
    const [socketStatus, setSocketStatus] = useState<ESocketConnection>(ESocketConnection.INITIAL);

    useEffect(()=>{
        const connected = SocketAPI.subscribeOnConnectionStatus(setSocketStatus);

        return connected.unsubscribe
    },[]);

    return socketStatus
}