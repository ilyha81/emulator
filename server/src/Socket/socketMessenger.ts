import {Timer} from '../Timer/interval';
import * as socketio  from "socket.io";
import * as core from "express-serve-static-core";
import {IOperationView} from "../Operation/operation";
import {Server as Socket} from 'socket.io';

/**
 * Типы сообщений.
 */
export enum ESocketMsg {
    MESSAGE = "MESSAGE",
    SHOW_DATA = "SHOW_DATA",
    TEST = "TEST",
}

/**
 * Проверка занятых портов.
 */
const ports: Set<number> = new Set;

export class SocketMessenger {
    socket: socketio.Server;

    static isPortUsed = (port: number) => ports.has(port);

    constructor (socket: Socket, getOperations: () => IOperationView[]) {
        this.socket = socket;

        this.socket.on('connection', (client: socketio.Socket) => {
            console.log('Клиент подключился');

            client.emit(ESocketMsg.MESSAGE, "ПОДКЛЮЧЕНИЕ УСПЕШНО !");
        });
    }

    /**
     * Обработка приема сообщений.
     * @param type
     * @param cb
     */
    on = (type: string, cb: (client: any) => any) => {
        this.socket.on(type, cb)
    };

    emitMsg = (type: string, msg: object) => {
        if (typeof msg === "string") {
            this.socket.emit(type, msg);
        } else {
            this.socket.emit(type, JSON.stringify(msg))
        }
    }
}
