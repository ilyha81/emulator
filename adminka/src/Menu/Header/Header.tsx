import * as React from "react";
import "./Header.css";
import logo from '../../logo.svg';
import {ESocketConnection} from "../../Socket/socket";

/**
 * Header props.
 * @prop {ESocketConnection} socketStatus Socket connection status.
 */
interface IHeader {
    socketStatus: ESocketConnection
}

/**
 * Хэдер админки. Переключение страниц.
 */
export const Header: React.FunctionComponent<IHeader> = ({children, socketStatus}) => {
    let badge = 'badge-warning';
    switch (socketStatus) {
        case ESocketConnection.CONNECTED:
            badge = 'badge-success';
            break;
        case ESocketConnection.FAILED:
        case ESocketConnection.DISCONNECTED:
            badge = 'badge-danger';
            break;
    }

    return <div className="Header">
        Хедер
        <div className="float-right" style={{width: 160, height: 100, overflow: 'hidden'}}>
            <img placeholder="HeaderLogo" id="HeaderLogo" src={logo}
                 className={`logo ${socketStatus === ESocketConnection.CONNECTED ? 'logo-rotation' : ''}`}
                 alt="HeaderLogo"/>
            <div><span className="badge badge-dark">Socket: </span><span
                className={`badge ${badge}`}>{socketStatus}</span></div>
        </div>
        {children}
    </div>
};