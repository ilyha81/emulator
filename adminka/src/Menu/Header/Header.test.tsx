import * as React from 'react';
import {cleanup, render, screen} from '@testing-library/react';

import {Header} from './Header';
import {ESocketConnection} from "../../Socket/socket";

beforeEach(()=> {
   cleanup();
});

describe('testing HEADER', () => {


    test('Header is rendered successfully', () => {
        const {container} = render(<Header socketStatus={ESocketConnection.INITIAL}/>);

        expect(container).toMatchSnapshot();
        expect(container.querySelector('#HeaderLogo')).toBeInTheDocument();
    });

    test('Children are Drawing in Header!', () => {
        const VASYA = 'VASYA';
        const PETYA = 'PETYA';
        const Child: React.FunctionComponent<{str: string}> = ({str}) => {
            return <div>{str}</div>
        };

        render(
            <Header socketStatus={ESocketConnection.INITIAL}>
                <Child key={VASYA} str={VASYA}/>
                <Child key={PETYA} str={PETYA}/>
            </Header>);
        screen.getByText(VASYA);
        screen.getByText(PETYA);
    });

    test('test BASE rendering of right Logo and Badge', () => {
       render(<Header socketStatus={ESocketConnection.INITIAL}/>);

        screen.getByText('Socket:');
        expect(screen.queryByText('Socket:')).toHaveClass('badge', 'badge-dark');
        expect(screen.queryByPlaceholderText('HeaderLogo')).toHaveClass('logo');
    });

    describe('Header displays socket connection status', () => {

        test('Testing INITIAL status', () => {
            render(<Header socketStatus={ESocketConnection.INITIAL} />);

            screen.getByText(ESocketConnection.INITIAL);
            expect(screen.queryByText(ESocketConnection.INITIAL)).toHaveClass('badge', 'badge-warning');
        });

        test('Testing CONNECTED status', () => {
            render(<Header socketStatus={ESocketConnection.CONNECTED} />);

            screen.getByText(ESocketConnection.CONNECTED);
            expect(screen.queryByText(ESocketConnection.CONNECTED)).toHaveClass('badge', 'badge-success');
            expect(screen.queryByPlaceholderText('HeaderLogo')).toHaveClass('logo-rotation');

        });

        test('Testing FAILED status', () => {
            render(<Header socketStatus={ESocketConnection.FAILED} />);

            screen.getByText(ESocketConnection.FAILED);
            expect(screen.queryByText(ESocketConnection.FAILED)).toHaveClass('badge','badge-danger');
        });

        test('Testing DISCONNECTED status', () => {
            render(<Header socketStatus={ESocketConnection.DISCONNECTED} />);

            screen.getByText(ESocketConnection.DISCONNECTED);
            expect(screen.queryByText(ESocketConnection.DISCONNECTED)).toHaveClass('badge', 'badge-danger');
        })

    });

});