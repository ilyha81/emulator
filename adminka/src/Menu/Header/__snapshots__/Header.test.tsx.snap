// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing HEADER Header is rendered successfully 1`] = `
<div>
  <div
    class="Header"
  >
    Хедер
    <div
      class="float-right"
      style="width: 160px; height: 100px; overflow: hidden;"
    >
      <img
        alt="HeaderLogo"
        class="logo "
        id="HeaderLogo"
        placeholder="HeaderLogo"
        src="logo.svg"
      />
      <div>
        <span
          class="badge badge-dark"
        >
          Socket: 
        </span>
        <span
          class="badge badge-warning"
        >
          INITIAL
        </span>
      </div>
    </div>
  </div>
</div>
`;
